### DISCUSSION TITLE
Is it possbile to call a function in nodejs during e2e tests?

### DISCUSSION THREAD

**Discussion by leilengma** [cypress-io/cypress#30166] [Questions and Help]
Hi, i am very new to cypress and still playing around with it.
In my case I need to test the pairing function from a website to an iot device.

So for me it is necessary to simulate a device in nodejs to send the correction information to the webservice when user clicks on "pair" button on the website. 

I know there is a beforeEach hook, but it is already in the browser when this hook is getting called. I really need the node enviroment to support the azure-iot-device library.

So I would like to ask if there is a way to jump out of the browser and execute some code in node during the test execution?

My current idea is to start a server before test execution and provide apis to perform all necessary actions via calling apis.

Thanks a lot in advance!

**Comment by portyu9** [comment]
Hi @leilengma 

Yes. `cy.task()` is the Cypress API intended for executing Node-side code from an E2E test.

You can register your IoT-device operation in `setupNodeEvents`:

```ts
// cypress.config.ts
import { defineConfig } from 'cypress'

export default defineConfig({
  e2e: {
    setupNodeEvents(on) {
      on('task', {
        async pairDevice(options) {
          // Use azure-iot-device here.
          // await the required device/service operation.

          return null
        },
      })
    },
  },
})
```

Then call it from the test:

```ts
cy.get('[data-cy=pair-device]').click()

cy.task('pairDevice', {
  deviceId: 'device-123',
})

cy.contains('Paired').should('be.visible')
```

A `cy.task()` may return a Promise, and Cypress will wait for it to resolve. The arguments/result should be serializable, and the task must eventually finish.

If your simulated IoT device needs a **long-running connection/listener**, do not put a never-ending server loop inside one `cy.task()`. Start that simulator as part of the Node-side test lifecycle or as an external process, then expose short tasks to control it, send messages, query its state, and clean it up.

That gives you the browser/Node split you are looking for without needing to expose your own HTTP API solely for Cypress.

No screenshots or attachments were included in this discussion.

Please mark this answer as accepted if it helped, thank you.
